lib/pathanalysis.cpp

```diff
@@ -166,8 +166,8 @@ PathAnalysis::Progress PathAnalysis::forwardRange(const Token* startToken, const
             if (f(info) == Progress::Break)
                 return Progress::Break;
         }
-        // Prevent infinite recursion
-        if (tok->next() == start)
+        // Try to prevent range overflow and infinite recursion
+        if (tok == endToken || tok->next() == start)
```

Comment:
It would be better to use `precedes(tok, endToken)` as the stopping condition in the for loop instead. Similar to how it is done here: 

https://github.com/danmar/cppcheck/blob/main/lib/forwardanalyzer.cpp#L552

Reply:
That sure could be an option. At least for the case my change was applied. But I still wonder what would happen if the endToken is the last token, and tok->next() would put you on the files first token. Which the original line was for.

For correct code in the token list this is not a big risk. But for odd/strange/faulty code, I have to little experience what kind of problem these tokAt() and linkAt() can create.

Reply:
Even in the code you link to there is an extra check for the cyclic case.

Reply:
I like the idea that all 'similar' loops use the same checks. Only thing I have against it is that it adds CPU cost with a function call. But hopefully it adds safety.

Reply:
Another question is. Should I also change the tok->next() to the same check in the loop you showed. And also throw an exception?